File: backend/app/schemas/plan.py
"""Study-plan contracts.

Like the other schema modules, these Pydantic models are the source of truth for
the OpenAPI schema, which generates the extension's TS types
(``extension/src/lib/types.ts`` — never hand-edited).
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

PlanItemKind = Literal["uncovered_topic"]

# Cap on uncovered topics in the plan. V2 scope: a resequencing of
# already-computed gaps, not a new generator.
MAX_UNCOVERED_TOPICS = 5


class PlanItem(BaseModel):
    kind: PlanItemKind
    title: str
    rationale: str


class StudyPlanResponse(BaseModel):
    space_id: UUID
    items: list[PlanItem]
    # When this composition ran — there is nothing cached to be stale, so this
    # is just "now," not a generation timestamp like coverage's.
    generated_at: datetime

File: frontend/src/lib/plan.ts
/**
 * Study plan, as returned by the backend.
 *
 * Shapes mirror `StudyPlanResponse` in `backend/app/schemas/plan.py`.
 * Hand-written per feature, following `lib/coverage.ts` — there is no
 * generated database types file.
 */
import { apiFetch } from '@/lib/api'

export type PlanItemKind = 'uncovered_topic'

export interface PlanItem {
  kind: PlanItemKind
  title: string
  rationale: string
}

export interface StudyPlanResponse {
  space_id: string
  items: PlanItem[]
  generated_at: string
}

export async function getStudyPlan(token: string, spaceId: string): Promise<PlanItem[]> {
  try {
    const res = await apiFetch(`/v1/spaces/${spaceId}/plan`, {}, token)
    if (!res.ok) return []
    return ((await res.json()) as StudyPlanResponse).items
  } catch (error) {
    console.error('Failed to load the study plan:', error)
    return []
  }
}
